### PR TITLE
Eliminate cn_node_loc, use node ID to rehydrate nodes

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

**Comment by beckerg** [comment]
This patch elimiantes cn_node_loc and we now use node ID to rebuild the tree at kvdb open.
Probably Alex and Gaurav should heavily scrutinize this patch.
All tests pass, as does repeated application of kmt.  ycsb native seems ok so far, will run mongod in the interim.

**Review by tristan957 — APPROVED** [review]

File: tests/unit/cn/cn_tree_test.c [hunk 1]
```diff
@@ -536,10 +528,10 @@ test_init(struct test *t, struct test_params *params, struct mtf_test_info *lcl_
     t->mtf = lcl_ti;
 }
 
-static u32
-num_kvsets_in_node(u32 node_level, u32 node_offset)
+static uint64_t
+num_kvsets_in_node(uint64_t nodeid)
 {
-    return 1 + node_offset % 3;
+    return 1 + (nodeid % 4);
```
Comment (beckerg) [inline tests/unit/cn/cn_tree_test.c:534]: Ooops, was mod 3 there for any particular reason?  Seems to work ok with mod 4...

**Comment by beckerg** [comment]
checkoss failed again, so much output, no obvious error message.  Guess I'll take a look at it...

**Comment by beckerg** [comment]
Whoa, I really wish I hadn't looked at that list of offensive words...

**Comment by tristan957** [comment]
I wish we just used common sense to check for words. We don't need to be pinging MIT on every PR.

**Comment by beckerg** [comment]
Probably pulling it once a day and caching it on /share or something would be sufficient, I really doubt it gets updated very often...

**Comment by tristan957** [comment]
That isn't how the github runners work

**Comment by beckerg** [comment]
IMP load running fine, will take an hour to see how things go when the ID space wraps...

File: lib/cn/cn_tree.c [hunk 2]
```diff
@@ -2276,18 +2265,16 @@ cn_tree_fini(void)
...
 struct cn_tree_node *
-cn_tree_find_node(struct cn_tree *tree, const struct cn_node_loc *loc)
+cn_tree_find_node(struct cn_tree *tree, uint64_t nodeid)
 {
     struct cn_tree_node *tn;
     struct tree_iter iter;
 
     tree_iter_init(tree, &iter);
 
     while (NULL != (tn = tree_iter_next(tree, &iter))) {
```
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/cn/cn_tree.c:2275]: Cursors and Spills have to iterate using route map.  Might as well do the same for diagnostics so nodes could be printed in order from left to right.

[the remark this answers is not in the document]
Comment (alexttx) [reply lib/cn/cn_tree.c:2275]: We could eventually rework open logic to create nodes in a private list of node structs, then find their max keys, then install them into the route map.  Then we wouldn't need this `cn_tree_find_node` api.

[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/cn_tree.c:2275]: Yeah, I think there's a lot of follow-on cleanup to be done before we get there.  Would prefer having only one way to do things vs several.  It was jarring with this patch that nodes now come out via the rest tree dump in reverse node ID order, except for the root node which comes first.  I'm not clear on why that happens, but I'll get there eventually...

[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/cn_tree.c:2275]: I'll open a ticket to try and fix this cn-tree-find-node issue...

File: tests/unit/cn/cn_tree_test.c (hunk 1 above)
Comment (alexttx) [reply tests/unit/cn/cn_tree_test.c:534]: any leaf node should work.  the mocked tree used to have 0,1,2 kvsets in cycle (or something like that).  and the mod 3 was to get a node with > 0 kvsets.  I changed to mock to have 3,4,5 in a cycle.  so it should no longer matter which leaf you get.

**Comment by beckerg** [comment]
IMP load has now circled back around, hasn't tripped any asserts, perf is good, yay!

**Pull request merged**